apps/asap-server/src/handlers/webhooks/webhook-run-migrations.ts

```diff
@@ -0,0 +1,249 @@
+import { migrationDir } from '../../config';
+import pinoLogger from '../../utils/logger';
+
+const squidexClient = new Squidex<RestMigration>('migrations');
+
+export const runFactory = (
+  logger: Logger,
+  client: Squidex<RestMigration>,
+  readDir: typeof fsPromise.readdir,
+  importModule: ImportModuleFromPath,
+): Handler => async () => {
+  const getMigrationPaths = getMigrationPathsFromDirectoryFactory(readDir);
+  const filterUnexecutedMigrations = filterUnexecutedMigrationsFactory(client);
+  const saveExecutedMigration = saveExecutedMigrationFactory(client);
+  const getMigrationsFromPaths = getMigrationsFromPathsFactory(
+    importModule,
+    logger,
+  );
+
+  const migrationPaths = await getMigrationPaths();
+  const unexecutedMigrationPaths = await filterUnexecutedMigrations(
+    migrationPaths,
+  );
+
+  logger.debug(
+    { migrations: unexecutedMigrationPaths },
+    'Outstanding migrations',
+  );
+
+  const migrations = await getMigrationsFromPaths(unexecutedMigrationPaths);
+
+  const executedMigrations: string[] = [];
+
+  for (const migration of migrations) {
+    logger.debug(`Executing migration '${migration.getPath()}`);
+    try {
+      await migration.up();
+      executedMigrations.push(migration.getPath());
+      logger.info(`Executed migration '${migration.getPath()}`);
+
+      await saveExecutedMigration(migration.getPath());
+      logger.debug(`Saved migration progress`);
+    } catch (error) {
+      logger.error(
+        error,
+        `Error executing the migration ${migration.getPath()}`,
+      );
+      break;
+    }
+  }
+
+  logger.info(`Executed and saved ${executedMigrations.length} migrations`);
+};
+
+export const rollbackFactory = (
+  logger: Logger,
+  client: Squidex<RestMigration>,
+  importModule: ImportModuleFromPath,
+): Handler => async () => {
+  const getLatestMigrationPathFromDb = getLatestMigrationPathFromDbFactory(
+    client,
+  );
+  const getMigrationsFromPaths = getMigrationsFromPathsFactory(
+    importModule,
+    logger,
+  );
+  const removeExecutedMigration = removeExecutedMigrationFactory(client);
+
+  const migrationPath = await getLatestMigrationPathFromDb();
+
+  logger.debug({ migrations: migrationPath }, 'Latest migrations');
+
+  if (migrationPath !== null) {
+    let migration: Migration;
+    try {
+      [migration] = await getMigrationsFromPaths([migrationPath]);
+    } catch (error) {
+      logger.error(
+        error,
+        `Could not load the migration from file ${migrationPath}`,
+      );
+
+      return;
+    }
+
+    logger.debug(`Executing rollback of migration '${migration.getPath()}`);
+
+    try {
+      await migration.down();
+    } catch (error) {
+      logger.error(
+        error,
+        `Error executing the rollback of migration ${migration.getPath()}`,
+      );
+      return;
+    }
+
+    logger.debug('Finished executing rollback, saving progress...');
+
+    await removeExecutedMigration(migrationPath);
```

Comment:
We should move the delete statement inside the try/catch. There's not error handling for this function

Reply:
On the same concerns me in the initial setup for each function:
- Rollback
```
  const getLatestMigrationPathFromDb = getLatestMigrationPathFromDbFactory(
    client,
  );
  const getMigrationsFromPaths = getMigrationsFromPathsFactory(
    importModule,
    logger,
  );
  const removeExecutedMigration = removeExecutedMigrationFactory(client);

  const migrationPath = await getLatestMigrationPathFromDb();
```

- Run:
```
  const getMigrationPaths = getMigrationPathsFromDirectoryFactory(readDir);
  const filterUnexecutedMigrations = filterUnexecutedMigrationsFactory(client);
  const saveExecutedMigration = saveExecutedMigrationFactory(client);
  const getMigrationsFromPaths = getMigrationsFromPathsFactory(
    importModule,
    logger,
  );

  const migrationPaths = await getMigrationPaths();
  const unexecutedMigrationPaths = await filterUnexecutedMigrations(
    migrationPaths,
  );
```
All this IO is not error handled. What do you think?

Reply:
the factory building is obviously not doing any work so i dont think it needs to be try catched, with regards to the reading of paths - again i think if this goes wrong and fails i want the lambda to fail. Ideally we would have some sensible alerting in place but since we run it manually ourselves it will just go red which is desired. Or do you mean i should just try catch it and throw something a little more meaningful?

Reply:
Maybe we can do with native logging for the setup bit, we want it to crash nonetheless. Let's just handle the `delete` call. 😉